src/broker/test/overlay.c

```diff
@@ -160,56 +206,44 @@ void single (flux_t *h)
     ctx_destroy (ctx);
 }
 
-int duo_init_cb (struct overlay *ov, void *arg)
+void recv_cb (const flux_msg_t *msg, overlay_where_t from, void *arg)
 {
     struct context *ctx = arg;
 
-    flux_log (ctx->h, LOG_INFO, "duo_init_cb called for %s", ctx->name);
-    return 0;
-}
-
-void socket_cb (struct overlay *ov, void *arg)
-{
-    struct context *ctx = arg;
+    diag ("%s message received",
+          from == OVERLAY_UPSTREAM ? "upstream" : "downstream");
+    ctx->msg = flux_msg_incref (msg);
     flux_reactor_stop (flux_get_reactor (ctx->h));
 }
 
 void timeout_cb (flux_reactor_t *r, flux_watcher_t *w, int revents, void *arg)
 {
+    diag ("receive timeout");
     errno = ETIMEDOUT;
     flux_reactor_stop_error (r);
 }
 
-flux_msg_t *recvmsg_child_timeout (struct context *ctx, double timeout)
+const flux_msg_t *recvmsg_timeout (struct context *ctx, double timeout)
 {
-    flux_reactor_t *r = flux_get_reactor (ctx->h);
-    flux_watcher_t *w;
-
-    overlay_set_child_cb (ctx->ov, socket_cb, ctx);
-    if (!(w = flux_timer_watcher_create (r, timeout, 0., timeout_cb, ctx)))
-        BAIL_OUT ("flux_timer_watcher_create failed");
-    flux_watcher_start (w);
 
-    if (flux_reactor_run (r, 0) < 0)
-        return NULL;
-    return overlay_recvmsg_child (ctx->ov);
-}
-
-flux_msg_t *recvmsg_parent_timeout (struct context *ctx, double timeout)
-{
     flux_reactor_t *r = flux_get_reactor (ctx->h);
     flux_watcher_t *w;
 
-    overlay_set_parent_cb (ctx->ov, socket_cb, ctx);
+    flux_msg_decref (ctx->msg);
+    ctx->msg = NULL;
+
     if (!(w = flux_timer_watcher_create (r, timeout, 0., timeout_cb, ctx)))
         BAIL_OUT ("flux_timer_watcher_create failed");
     flux_watcher_start (w);
 
     if (flux_reactor_run (r, 0) < 0)
-        return NULL;
-    return overlay_recvmsg_parent (ctx->ov);
+        return NULL; // timeout
+    return ctx->msg;
```

Comment:
watcher `w` leaked here, but that wasn't introduced in this PR so feel free to ignore it.

Reply:
Oh thanks!  I'll fix that in the next PR that I have cooking for this area of code.  I'll go ahead and set MWP.